src/util/simplify_expr_int.cpp

```diff
@@ -848,32 +856,43 @@ bool simplify_exprt::simplify_bitwise(exprt &expr)
 
 bool simplify_exprt::simplify_extractbit(exprt &expr)
 {
-  const typet &op0_type=expr.op0().type();
-
-  if(!is_bitvector_type(op0_type))
-    return true;
+  PRECONDITION(expr.id() == ID_extractbit);
+  auto &extractbit_expr = to_extractbit_expr(expr);
```

Comment:
Pass `extractbit_exprt` as parameter, and let `to_extractbit_expr` do all the work (at the call site) - no need for duplication.

Reply:
This can't take an `extractbit_exprt` because the parameter is an in/out parameter, and the result is not always going to be another `extractbit_exprt`

Reply:
We have this problem across all simplifier functions. The cleanest approach might be changing the return type from `bool` to something more elaborate so that we can return the modified expression, and would instead take a const reference as argument. Thus far, the workaround was using `swap` instead of direct assignments. I think it's important to get rid of the duplicate checks on input parameters (there should only be a check at the call site).

Reply:
@tautschnig I'm going to split this part out - and make a note for it to be considered in the context of #2939 and your remarks on #2910 (comment)